src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -92,12 +92,41 @@ def upgrade_status(self) -> orchestrator.UpgradeStatusSpec:
         if self.upgrade_state:
             r.target_image = self.target_image
             r.in_progress = True
+            r.progress, r.services_complete = self._get_upgrade_info()
+            # accessing self.upgrade_info_str will throw an exception if it
+            # has not been set in _do_upgrade yet
+            try:
+                r.message = self.upgrade_info_str
+            except AttributeError:
+                pass
             if self.upgrade_state.error:
                 r.message = 'Error: ' + self.upgrade_state.error
             elif self.upgrade_state.paused:
                 r.message = 'Upgrade paused'
         return r
 
+    def _get_upgrade_info(self) -> Tuple[str, List[str]]:
+        if not self.upgrade_state or not self.upgrade_state.target_digests:
+            return '', []
+        daemons = [d for d in self.mgr.cache.get_daemons() if d.daemon_type in CEPH_UPGRADE_ORDER]
+        done = 0  # total number of upgraded daemons
+        old = 0  # number of seen upgraded daemons before checking current service
+        completed = []
+        for daemon_type in CEPH_UPGRADE_ORDER:
+            count = 0  # number of daemons of this service type we've checked
+            for d in daemons:
+                if d.daemon_type != daemon_type:
+                    continue
+                if not d.container_image_digests:
+                    return '', []
+                count += 1
+                if any(d in self.upgrade_state.target_digests for d in (d.container_image_digests or [])):
+                    done += 1
+            if (done - old) == count and count != 0:
+                completed.append(daemon_type)
+            old = done
+        return '%s/%s daemons upgraded' % (done, len(daemons)), completed
```

Comment:
In my opinion is more clear the Adam's code.

Reply:
@sebastian-philipp This, slightly adjusted (mostly to avoid syntax and typing errors + change to completed_types line to make sure all daemons of that type are upgraded) version of what you posted works (tested on an upgrade with mgr, mon, crash and osds daemons). If you're happy with this I'm happy with swapping to it. It's definitely cleaner than what i originally did.

 ```
if not self.upgrade_state or not self.upgrade_state.target_digests:
        return '', []

daemons = [d for d in self.mgr.cache.get_daemons() if d.daemon_type in CEPH_UPGRADE_ORDER]

 if any(not d.container_image_digests for d in daemons if d.daemon_type == 'mgr'):
        return '', []

completed_daemons = [(d.daemon_type, any(d in self.upgrade_state.target_digests for d in (
        d.container_image_digests or []))) for d in daemons if d.daemon_type]

 done = len([True for completion in completed_daemons if completion[1]])

 completed_types = list(set([completion[0] for completion in completed_daemons if all(
         c[1] for c in completed_daemons if c[0] == completion[0])]))

return '%s/%s ceph daemons upgraded' % (done, len(daemons)), completed_types

```

Reply:
However you want. 

Your original code has simpler statements but the overall logic is more quirky. 
My alternative has super complicated statements, but the control flow is super simple. 

:shrug: